Push notifications for PWA

jdiehl opened the pull request:
@ConnorLand

jdiehl commented:
Needs review from @allard-mx

ConnorLand commented:
@jdiehl and I need to duplicate these changes in Mx9, right?

allard-mx commented on content/en/docs/refguide/mobile/using-mobile-capabilities/push-notifications/notif-implement-pwa.md:
```diff
@@ -0,0 +1,61 @@
+
+## 2 Initialize Firebase
+
+Push notifications in progressive web apps require firebase to be set up as early as possible. To do so, you need to create a custom index.html and initialize firebase there. Follow these steps to do so:
```
Firebase is capitalized in other parts of the module docs, would prefer using a capital here as well (and the other mention in the second sentence)

allard-mx commented on content/en/docs/refguide/mobile/using-mobile-capabilities/push-notifications/notif-implement-pwa.md:
````diff
@@ -0,0 +1,61 @@
+const messaging = firebase.messaging();
+```
+
+5. Back in Studio Pro, set the constant `WebPushVapidKey` found in `_USE ME/Web/` in the Push Notifications module to the web certificate from part 3.
````
I would change `web certificate` to `public key of the Web Push certificate you created in part 3`

allard-mx commented on content/en/docs/refguide/mobile/using-mobile-capabilities/push-notifications/notif-send-test.md:
```diff
@@ -44,3 +44,6 @@ If you did not receive the message, check the application's logs to see if the m
 | Sending a message causes a **Request contains an invalid argument** error or **400: Bad Request**. | Your **Project ID** does not match the **project_id** in your private key *json* file. | Upload the correct file or [generate a new private key](/refguide/mobile/using-mobile-capabilities/push-notifications/setting-up-google-firebase-cloud-messaging-server/#setting-up-a-service-account) in Fireb ...
 | Mendix Runtime exception on JavaAction 'DecryptString': **Key should not be empty**. | This module depends on the **Encryption** module, which requires a key. | Set the constant **EncryptionKey** in the **Encryption** module with a key of exactly 16 characters. |
 | Error sending message: **Error reading credentials from stream, 'type' field not specified. at PushNotifications.SendFCMMessages (JavaAction : 'GetFCMAccessToken')**. | The wrong private key file was uploaded. | Upload the correct file or [generate a new private key](/refguide/mobile/using-mobile-capabilities/push-notifications/setting-up-google-firebase-cloud-messaging-server/#setting-up-a-ser ...
+| Not receiving push notifications on a PWA | Push notifications on PWAs are only shown when the app is not open | Close the app and try again |
```
Maybe change this in something like `not open on the foreground`?

allard-mx commented on content/en/docs/refguide/mobile/using-mobile-capabilities/push-notifications/notif-send-test.md:
```diff
@@ -44,3 +44,6 @@ If you did not receive the message, check the application's logs to see if the m
 | Mendix Runtime exception on JavaAction 'DecryptString': **Key should not be empty**. | This module depends on the **Encryption** module, which requires a key. | Set the constant **EncryptionKey** in the **Encryption** module with a key of exactly 16 characters. |
 | Error sending message: **Error reading credentials from stream, 'type' field not specified. at PushNotifications.SendFCMMessages (JavaAction : 'GetFCMAccessToken')**. | The wrong private key file was uploaded. | Upload the correct file or [generate a new private key](/refguide/mobile/using-mobile-capabilities/push-notifications/setting-up-google-firebase-cloud-messaging-server/#setting-up-a-ser ...
+| Not receiving push notifications on a PWA | Push notifications on PWAs are only shown when the app is not open | Close the app and try again |
+| Not receiving push notifications on a PWA on iOS | iOS PWAs require https to run | Deploy your app and try again |
```
We could mention that you can use a tool like ngrok during local development

jdiehl replied:
I'd rather not - I don't want to maintain references to external sites that might change or remove their service at any time. Developers can test web push notifications easily in Chrome.

allard-mx commented on content/en/docs/refguide/mobile/using-mobile-capabilities/push-notifications/notif-send-test.md:
```diff
@@ -44,3 +44,6 @@ If you did not receive the message, check the application's logs to see if the m
 | Error sending message: **Error reading credentials from stream, 'type' field not specified. at PushNotifications.SendFCMMessages (JavaAction : 'GetFCMAccessToken')**. | The wrong private key file was uploaded. | Upload the correct file or [generate a new private key](/refguide/mobile/using-mobile-capabilities/push-notifications/setting-up-google-firebase-cloud-messaging-server/#setting-up-a-ser ...
+| Not receiving push notifications on a PWA | Push notifications on PWAs are only shown when the app is not open | Close the app and try again |
+| Not receiving push notifications on a PWA on iOS | iOS PWAs require https to run | Deploy your app and try again |
+| Clicking on a push notification does not open the PWA | Push notifications for PWAs require a WebLink to be set. | Make sure the WebLink is set to a valid path of your application (for example "/") |
```
I would change `WebLink` into `Web link`, since this is how it's shown in the send message pop-up

allard-mx commented on content/en/docs/refguide/mobile/using-mobile-capabilities/push-notifications/setting-up-google-firebase-cloud-messaging-server.md:
```diff
@@ -122,3 +122,21 @@ The list at the bottom shows the Android and iOS applications that you have conf
+
+## 9 Set up Web Push Certificate
+
+For progressive web apps, an additional certificate is needed. You can skip this step for native mobile apps. To obtain the certificate, again click the cogwheel on the upper-left side of the screen and select **Project settings**. Then navigate to the **Cloud Messaging** tab and click on **Generate Key Pair**:
```
I would change `certificate` into `public key`, since it is only the public part of the certificate. Using the word `certificate` might raise concerns, since the constant is exposed to the client in the app.

allard-mx reviewed (changes requested):
Looks very good to me! I left a few minor remarks.

jdiehl commented:
> @jdiehl and I need to duplicate these changes in Mx9, right?

Yes. The module will be compatible with 9.24. The setup is the same for mx 9

jdiehl commented:
> Looks very good to me! I left a few minor remarks.

@ConnorLand Allard approved the PR. Please align with him because the docs update should be in sync with a new marketplace release

Commit:
adressed last comment

Commit:
restructure parts 5 & 6

Commit:
part 4

Commit:
Addressed comments

Commit:
Up to part 3

Commit:
added implementing pwa

Commit:
Added troubleshooting

Commit:
Clarified iOS requirements

Commit:
added warning about script tag order

ConnorLand committed:
Format and edit

ConnorLand committed:
Duplicate in Mx9

ConnorLand committed:
Fix indentations

ConnorLand committed:
Duplicate changes in 9

ConnorLand committed:
Duplicate changes to Mx9

allard-mx commented on content/en/docs/refguide9/mobile/using-mobile-capabilities/push-notifications/notif-add-module-depends.md:
```diff
@@ -44,7 +44,7 @@ Skip this section if the [Nanoflow Commons](https://marketplace.mendix.com/link/
 ## 5 Implementing the Native Mobile Resources Module
 
-Skip this section if the [Native Mobile Resources](https://marketplace.mendix.com/link/component/109513/Mendix/Native-Mobile-Resources) module is already implemented in your app, or if you do not plan on using using local or native push notifications. Implement this module by doing the following:
+Skip this section if the [Native Mobile Resources](https://marketplace.mendix.com/link/component/109513/Mendix/Native-Mobile-Resources) module is already implemented in your app. This module is required even for progressive web apps. Implement this module by doing the following:
```
The change for Nanoflow Commons (one section above this one) seems to be missing here, where the `or if you do not plan on using using local or native push notifications` part got removed

ConnorLand replied:
Added in this section

MxMurshed replied:
It's also removed for **- 4 Implementing the Nanoflow commons module** line 37. `or if you do not plan on using using local or native push notifications` needs to be removed from there as well.

Note: it's removed for Mx10

MxMurshed replied:
@ConnorLand

ConnorLand committed:
Revert native 9 doc

allard-mx commented on content/en/docs/refguide9/mobile/using-mobile-capabilities/push-notifications/notif-build-native.md:
```diff
@@ -32,4 +32,4 @@ Now that you have implemented push notifications, it is time to build and deploy
 When building for local development, keep in mind that Mendix's Make it Native does not support push notifications. In order to use and test push notifications, you will have to build your own native app as described above and distribute it to an emulator (Android only) or test device.
 
-Now you are able to build, the next step is to run your app in an emulator or test device. This will let you send your first test push notification in the next section.
+Now you are able to build, the next step is to run your app in an emulator or test device. This will let you send your first test push notification in the next section.
```
The `notif-build-native.md` file seems to be deleted for the Mendix 10 refguide (see [this](https://github.com/jdiehl/docs/commit/4ba05aa394ff889603d5ec04be77703b8e3b0d32) commit), however, it seems the introduction still refers to it? The refguides for Mx 9 and 10 are not in sync now as far as I can see. @jdiehl Do you know why the file was deleted?

allard-mx replied:
Never mind, it seems the part for building the native app has been moved into the `notif-implement-native.md` file (section 5, see my comment below). I do think there are still some broken links though, in the index/introduction and in `notif-implement-native.md` itself

ConnorLand replied:
Fixed here: https://github.com/mendix/docs/pull/8043/commits/6cd389fcbcb5457d6435e19f0feb4bec39b74f80

ConnorLand replied:
Sorry, actually finally fixed (with the Mx9 doc deleted as the Mx10 one is) here: https://github.com/mendix/docs/pull/8043/commits/fd92e20317edf11126bad1b378252dbd9a539d58

allard-mx commented on content/en/docs/refguide9/mobile/using-mobile-capabilities/push-notifications/notif-implement-pwa.md:
```diff
@@ -0,0 +1,61 @@
+---
+title: "Part 7: Implement Push Notifications in Your Progressive Web App"
```
Title mentions part 7, whilst the link title below says part 6, this seems incorrect to me? In `notif-config-push.md` this page gets referred to as part 6.

MxMurshed replied:
The title is still referring to part 7 whilst the link is to part 6.

Note: The Mx10 refguide is referring to part 6 for both.

MxMurshed replied:
I think this should be part 6 and the parts afterwards should be reduced by 1?

I found sections (1-2-3-4-5-**7-8-9**) -> missing section 6 and incremented 7,8.9?

allard-mx commented on content/en/docs/refguide9/mobile/using-mobile-capabilities/push-notifications/notif-send-test.md:
```diff
@@ -1,6 +1,6 @@
 ---
-title: "Part 7: Send Your First Test Push Notification"
-linktitle: "7. Test Push Notification"
+title: "Part 8: Send Your First Test Push Notification"
```
Same here, seems like something happened with the part numbering, it is out of sync with the Mx 10 refguide

MxMurshed replied:
This should be section part 7 as well as title part 7

allard-mx commented on content/en/docs/refguide/mobile/using-mobile-capabilities/push-notifications/notif-implement-native.md:
```diff
@@ -54,3 +56,25 @@ To ensure push notification integration executes properly on your native app you
 This will ensure that the correct objects are synchronized to your native apps.
 
 Now that you have everything set up, it is time to deploy your native app. See [Build a Native Mobile App with Push Notifications](/refguide/mobile/using-mobile-capabilities/push-notifications/notif-build-native/) for instructions on enabling push notifications when building a native app. 
```
Is this correct? It seems this part has been moved into the current file in section 5 below, so this link is broken?

ConnorLand replied:
Correct, therefore I've edited this paragraph.

ConnorLand committed:
Fix links

Now the links that point to the older deleted doc have been pointed to the new subsection

ConnorLand committed:
Delete Build Native in Mx9

ConnorLand committed:
Delete mention in parent doc

ConnorLand committed:
Address comment

ConnorLand committed:
Add link to Mx10 PN index page

ConnorLand committed:
Fix link errors

MendixMurshed commented on content/en/docs/refguide9/mobile/using-mobile-capabilities/push-notifications/notif-mult-devices.md:
This should be Part 8?

MxMurshed commented on static/attachments/howto/mobile/native-mobile/implementation/notifications/setting-up-google-firebase-cloud-messaging-server/push_notifications_web_config.png:
Is it ok to show the API key and the other project details? or should they be masked?

ConnorLand replied:
Masked this one with a blur effect.

MxMurshed commented on static/attachments/howto/mobile/native-mobile/implementation/notifications/setting-up-google-firebase-cloud-messaging-server/push_notifications_web_certificate2.png:
Shouldn't the key pair be masked?

ConnorLand replied:
Masked this one with a blur effect.

MxMurshed commented on static/attachments/howto/mobile/native-mobile/implementation/notifications/notif-implement-pwa/vapid-idwizard.png:
Shouldn't we mask the default value here as well?

ConnorLand replied:
Masked like Jim Carrey in the eponymous film.

ConnorLand committed:
Change number to 8

ConnorLand committed:
Set test doc to 7

ConnorLand committed:
Blur info

MxMurshed commented on content/en/docs/refguide9/mobile/using-mobile-capabilities/push-notifications/notif-add-module-depends.md:
Line 37: S`or if you do not plan on using using local or native push notifications` should be removed?

ConnorLand replied:
Fixed: https://github.com/mendix/docs/pull/8043/commits/ff44bc4faeb1750b0a076352eb86a1b8ec497de5

ConnorLand committed:
Remove optional language in Mx8

ConnorLand committed:
Delete confusing phrase in Mx8

ConnorLand committed:
Change part 7 to 6 in Ref9
